lib/nx_inflate.c

```diff
@@ -962,14 +962,25 @@ static int nx_inflate_(nx_streamp s, int flush)
 
 	assert(s->used_out == 0);
 
-	/* if s->avail_out and  s->avail_in is 0, return */
+	/* If there is no input (s->avail_in == 0 && s->used_in == 0) or there
+	   is no space for output (s->avail_out == 0), return. */
 	if (s->avail_out == 0 || (s->avail_in == 0 && s->used_in == 0)) return Z_OK;
 	if (s->used_out == 0 && s->avail_in == 0 && s->used_in == 0) return Z_OK;
 	/* we should flush all data to next_out here, s->used_out should be 0 */
 
 	/* used_in is the data amount waiting in fifo_in; avail_in is
-	   the data amount waiting in the user buffer next_in */
-	if (s->avail_in < nx_config.soft_copy_threshold && s->avail_out > 0) {
+	   the data amount waiting in the user buffer next_in.
+	   Cache the input in fifo_in until we have enough data in order to send
+	   to the NX GZIP accelerator.
+	   Avoid executing this code when avail_in is 0.  That means either the
+	   end of the stream or the end of the current request.  There is
+	   nothing to copy anyway.
+	   The following code is not just an optimization, it is also required
+	   by NX because it may refuse to start processing a stream if the input
+	   is not large enough.  */
+	if (s->avail_in > 0
+	    && (s->avail_in + s->used_in < nx_config.soft_copy_threshold)
+	    && s->avail_out > 0) {
```

Comment:
IIUC, if the user is decompressing a small stream (smaller than the threshold), and the entire data is available at once, it will need to call inflate twice, right?
1st call ) avail_in = LEN , then LEN bytes are copied to fifo_in and no output generated
2nd call ) avail_in = 0, then NX is triggered because the first condition is false then all input is decompressed.

I haven't tested, but it seems this extra call is needed even if `flush == Z_FINISH`. `nx_inflate` should be able to let the user skip this second call if `Z_FINISH` or other flush parameters are provided.

But I suspect there are other issues with the way libnxz handles these other flush modes on inflate, so this could be part of a separate PR.

Reply:
> it will need to call inflate twice, right?

I am guessing no.  If the entire data is available at once, then the last Deflate block in the stream will have the BFINAL=1 set.  It means this is the last block.  As soon as inflate() sees the EOB symbol in that block with BFINAL=1, then it should say this is the end of data.  But depending on the format it may also look for additional 4 to 8 bytes in the trailer after EOB. 
 i.e., the checksum (zlib and gzip formats) and ISIZE field (gzip only).

Reply:
As a follow-up, this was just a guess on my part.  It may be useful setting up a small test with examples/zpipe.c to understand the zlib behavior. 

But I think your comment is valid for deflate().  How does deflate know it's the end of input unless you give it a Z_FINISH?  So, again guessing, if you give input with Z_NO_FLUSH but then realize there is no more input, I think you can end the stream with a 0 size input and a Z_FINISH.  In the inflate() case though, the end is already marked so I suspect inflate will return Z_STREAM_END independent of your flush mode